pages/api/cnpj/v1/[cnpj].js

```diff
@@ -0,0 +1,20 @@
+import microCors from 'micro-cors';
+import { getCnpjData } from '../../../../services/cnpj';
+
+const cors = microCors();
+const CACHE_CONTROL_HEADER_VALUE =
+  'max-age=0, s-maxage=86400, stale-while-revalidate';
+
+async function cnpjData(request, response) {
+  response.setHeader('Cache-Control', CACHE_CONTROL_HEADER_VALUE);
+  try {
+    const result = await getCnpjData(request.query.cnpj);
+    response.status(result.status);
+    response.json(result.data);
+  } catch (error) {
+    response.status(500);
```

Comment:
Só tenho uma dúvida aqui

Sempre que der erro na requisição, o erro HTTP 500 será retornado (Internal server error)

Mas se um CNPJ não for encontrado, deveríamos retornar o erro 404 (Not Found), não?

Reply:
Eu sempre fico na dúvida se usar o 404 ou 204 (no content)... rs
Neste caso talvez eu retornasse até um 204 :-p

Reply:
Hum… então o `catch` não pega só erros de servidores (_HTTP Status 5xx_) mas erros de requisição (_HTTP Status 4xx_) também! Já aprendi algo novo sobre o Axios.

A _Minha Receita_ atualmente vai retornar 404, mas eu concordo total com o @RodriAndreotti: 204 é uma resposta semanticamente melhor e vou mudar isso (hoje ainda, espero) na API.

Reply:
> (hoje ainda, espero)

Feito ✅ https://github.com/cuducos/minha-receita/commit/e315185136e9bc800245ecf9dc78b3dc066f4373
Assim que a CI terminar, se tudo der certo, entra em produção.

Reply:
Pronto:

```
$  curl -i -X POST -d "cnpj=00000000000000" "https://minhareceita.org"
HTTP/2 204
…
```

Agora esse caso não cai no `catch` mais por ser da família 200… e ainda é uma API mais semântica: indica (por convenção das famílias de status de respostas HTTP) que o conteúdo não existe ao invés de sugerir erro na requisição ✨

Reply:
Mestre, entendo o ponto mas é uma falha do ponto de vista de quem está buscando.

Eu como cliente ao fazer um get nessa rota de algum recurso que não existe (por exemplo o cpnj 123) espero encontrar a msg que o recurso não existe, não?

🤔 

Obs: essa é uma boa discussão tbm pra gente pensar em padrões do restante da api (ainda ainda não ta padronizada hehehe) ^^

Reply:
Vamos a implementação de quem consome por exemplo:

Com 404:
```
try {
  const { data } = await axios.get('https://brasilapi.com.br/api/cnpj/v1/123')

 console.log('Deu certo, uhulll')
 return data
} catch (error) {
  console.log('nao foi possível encontrar o CNPJ')
  throw error
}
```

Com 204:
```
try {
  const { data, statusCode } = await axios.get('https://brasilapi.com.br/api/cnpj/v1/123')
  
  if (statusCode === 204) {
    throw new Error('cnpj_not_found')
  }

 console.log('Deu certo, uhulll')
 return data
} catch (error) {
  console.log('nao foi possível encontrar o CNPJ')
  throw error
}
```

Qual dos 2 faz mais sentido pra quem ta integrando? 😬

Reply:
> Eu como cliente ao fazer um get nessa rota de algum recurso que não existe (por exemplo o cpnj 123) espero encontrar a msg (de erro) que o recurso não existe, não?

Acho que não. Por dois motivos: (a) o exemplo `123` não cairia em 204, e (b) o exemplo `123` traria uma mensagem de erro.

**Como eu entendo do lado da _Minha Receita_**

O usuário esperaria uma mensagem de erro **se** tivesse um erro na requisição, o que tecnicamente não tem no caso de, por exemplo, `cnpj=00000000000000` — considerando que uma sequência de 14 zeros é um CNPJ válido (apesar de inexistente).

Como não tem erro, o HTTP status informa o usuário que a requisição está correta, mas não existe recurso (objeto, o CNPJ em si) a ser exibido. Semanticamente, é isso que o 204 comunica:

> The HTTP `204 No Content` success status response code indicates that **a request has succeeded**, but that the client doesn't need to navigate away from its current page.
> 
> Da [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/204), com grifo meu.

Já no teu examplo, o caso do CNPJ enviado ser `123`, como 123 **não** é um CNPJ válido, a _Minha Receita_ vai retornar HTTP status 400, pois um CNPJ inválido é um erro da requisição. Nesse caso, como existe erro, [a resposta inclui a mensagem de erro](https://docs.minhareceita.org/cliente/#post): `{"message": "CNPJ 123 inválido."}`.

Adotar o 204 no caso em que não existe erro na requisição me parece também algo que reduz o tráfego e processamento da _Minha Receita_ — o que ajuda a aguentar o tranco. E esse é o foco da API: escalabilidade.

Agora, outro papo é o design da API de vocês…

**Como modelarmos a interface do BrasilAPI** 

Aqui podemos capturar o 204 e darem o tratamento que preferirem... pelo design atual, sempre que o status for 204 dá para mudar o HTTP status para outra coisa (404 e traduzir 204 como erro, por exemplo) e incluir a mensagem que acharem melhor. Essa decisão é mais de vocês do que minha : )

Minha dúvida é: devo explicitamente criar uma resposta HTTP?

```javascript
if (result.status == 204) {
  response.status(404)
  response.json({ message: "CNPJ não encontrado."})
}
```

Ou se só usar um erro já resolve?

```javascript
if (result.status == 204) {
  throw new Error("CNPJ não encontrado.")
}
```

**TLDR** Posso criar um commit tratando o 204 e transformando ele em um 404 para a integração do BrasilAPI?

Reply:
Então, eu sempre me pego no mesmo questionamento seu, @lucianopf !
Mas, apesar, de a implementação com 404 ser mais simples num primeiro instante, eu meio que tendo a pensar que ele é mais comum ser usado quando um servidor não encontra o recurso que responde pelo request feito.

Trecho da wikipedia: 
> The website hosting server will typically generate a "404 Not Found" web page when a user attempts to follow a broken or dead link; hence the 404 error is one of the most recognizable errors encountered on the World Wide Web. 

Enquanto que o 204, no meu ver, diz que o servidor foi capaz de processar aquela requisição, mas não tem nada para retornar para ela.
Trecho

> 204 No Content
>     The server successfully processed the request, and is not returning any content.[13]

No meu ver o 404 pode gerar alguma confunção com um URL inexistente e o camarada não saber se o resultado é por uma url inexistente ou somente pelo recurso não ser encontrado, o que acham?

Reply:
Rascunhei aqui, compartilhando antes de _comitar_:

```javascript
import microCors from 'micro-cors';
import { getCnpjData } from '../../../../services/cnpj';

const cors = microCors();
const CACHE_CONTROL_HEADER_VALUE =
  'max-age=0, s-maxage=86400, stale-while-revalidate';

async function cnpjData(request, response) {
  response.setHeader('Cache-Control', CACHE_CONTROL_HEADER_VALUE);
  try {
    const result = await getCnpjData(request.query.cnpj);
    if (result.status === 204) {
      // the CNPJ is valid but does not exist
      response.status(404);
      response.json({ message: `CNPJ ${request.query.cnpj} não encontrado.` });
      return;
    }
    response.status(result.status);
    response.json(result.data);
  } catch (error) {
    response.status(500);
    response.json(error.response.data);
  }
}

export default cors(cnpjData);
```

Reply:
Fala @cuducos, acho que houve um erro de interpretação de vocês nisso de 204 ou 404.

### 204
O 204 (No Content) diz que a requisição deu certo, mas nada será retornado. Geralmente esse status code é utilizado em update de registros, você pode ler mais sobre isso nesses links:

https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Status/204
https://restfulapi.net/http-status-204-no-content/

Como pode ver, o **comum** é utilizar isso em `PUT`, e quando falamos de comum, estamos falando sobre o que a maioria das pessoas está acostumada

### 404
Já, o 404 (Not Found) deve ser utilizado quando buscamos por um recurso **específico**, mas ele não foi encontrado. Essa é a situação que essa rota se encaixa (até em seu provedor, minhareceira.org). Se eu buscar pelo CNPJ `00000000000000`, ele é "válido", mas não será encontrado. Ou seja, nesse caso devemos dizer que o **recurso não foi encontrado** (404).

Você pode ler mais nesses links:
https://softwareengineering.stackexchange.com/questions/203492/when-to-use-http-status-code-404-in-an-api
https://stackoverflow.com/questions/26845631/is-it-correct-to-return-404-when-a-rest-resource-is-not-found
https://mundipagg.zendesk.com/hc/pt-br/articles/360035107851-C%C3%B3digos-de-status-da-API-HTTP-Status-Code- (Nesse link troque `customer` por `cnpj`)

### 400

Sobre o erro 400, concordamos contigo, se a busca for pelo CNPJ `123`, deve ser retornado um erro de CNPJ inválido, com status code 400. Quando o @lucianopf citou o CNPJ `123`, era só um valor de exemplo, ele não quis entrar no mérito do valor ser válido ou não, apenas quis dizer que o CNPJ `123` não seria encontrado, logo um 404 deveria ser retornado.

TLDR: Colocar 204 como resposta no `minhareceita.org` foi um equivoco, seria muito melhor se você deixasse como 404, pela própria saúde do seu sistema. Os demais tratamentos de status conseguimos fazer aqui no brasilapi, sem problemas =]

Reply:
> Colocar 204 como resposta no minhareceita.org foi um equivoco

Discordo, mas isso não é assunto para BrasilAPI. De qualquer forma, de acordo com a implementação do _commit_ 96190bc33d1306dea2fe894084a6d587ce762ad2 aqui nesse PR, em ambos os casos (retornando 204 ou 404 na _Minha Receita_ para CNPJ válido mas inexistente) vai dar da mesma, não? Ou tem algo a mudar nesse PR? Isso que não entendi.

Reply:
> 
> 
> Fala @cuducos, acho que houve um erro de interpretação de vocês nisso de 204 ou 404.
> ### 204
> 
> O 204 (No Content) diz que a requisição deu certo, mas nada será retornado. Geralmente esse status code é utilizado em update de registros, você pode ler mais sobre isso nesses links:
> 
> https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Status/204
> https://restfulapi.net/http-status-204-no-content/
> 
> Como pode ver, o **comum** é utilizar isso em `PUT`, e quando falamos de comum, estamos falando sobre o que a maioria das pessoas está acostumada
> ### 404
> 
> Já, o 404 (Not Found) deve ser utilizado quando buscamos por um recurso **específico**, mas ele não foi encontrado. Essa é a situação que essa rota se encaixa (até em seu provedor, minhareceira.org). Se eu buscar pelo CNPJ `00000000000000`, ele é "válido", mas não será encontrado. Ou seja, nesse caso devemos dizer que o **recurso não foi encontrado** (404).
> 
> Você pode ler mais nesses links:
> https://softwareengineering.stackexchange.com/questions/203492/when-to-use-http-status-code-404-in-an-api
> https://stackoverflow.com/questions/26845631/is-it-correct-to-return-404-when-a-rest-resource-is-not-found
> https://mundipagg.zendesk.com/hc/pt-br/articles/360035107851-C%C3%B3digos-de-status-da-API-HTTP-Status-Code- (Nesse link troque `customer` por `cnpj`)
> ### 400
> 
> Sobre o erro 400, concordamos contigo, se a busca for pelo CNPJ `123`, deve ser retornado um erro de CNPJ inválido, com status code 400. Quando o @lucianopf citou o CNPJ `123`, era só um valor de exemplo, ele não quis entrar no mérito do valor ser válido ou não, apenas quis dizer que o CNPJ `123` não seria encontrado, logo um 404 deveria ser retornado.
> 
> TLDR: Colocar 204 como resposta no `minhareceita.org` foi um equivoco, seria muito melhor se você deixasse como 404, pela própria saúde do seu sistema. Os demais tratamentos de status conseguimos fazer aqui no brasilapi, sem problemas =]

Fala @murilohns!
Beleza?
Cara, estava relendo as especificações do RFC 7231 que fala sobre esses códigos de erro, e, além de outras informações,  achei isso:

>  A resposta 204 permite que um servidor indique que a ação foi aplicada com sucesso ao recurso de destino, ao mesmo tempo que implica que o agente do usuário não precisa se afastar de sua "visualização de documento" atual (se houver). O servidor assume que o agente do usuário fornecerá alguma indicação do sucesso ao seu usuário, de acordo com sua própria interface, e aplicará quaisquer metadados novos ou atualizados na resposta à sua representação ativa.

   Por exemplo, um código de status 204 é comumente usado com interfaces de edição de documentos correspondendo a uma ação "salvar", de modo que o documento sendo salvo permaneça disponível para o usuário editar. Também é freqüentemente usado com interfaces que esperam que as transferências de dados automatizadas sejam predominantes, como em sistemas de controle de versão distribuídos.

Tradução do google tradutor, estava com preguiça :-p ... rs

Basicamente, o que entendo dessa parte, é que o servidor processou a informação e não tem nada para retornar sobre ela, o que no meu ver, se encaixa também em uma requisição GET que não encontrou informações.

Porém devo admitir que também fala do PUT, como você comentou, e como também é comentado no link da Mozilla.
Porém a RFC cita o PUT como um exemplo onde nada precisa ser retornado.

Link: https://tools.ietf.org/html/rfc7231#section-6.3.5

O RFC 2616 diz isso:
Estava lendo algum tempo atrás uma discussão muito interessante à respeito no StackOverflow: https://stackoverflow.com/questions/34312023/http-get-request-status-204-vs-404

E na realidade, cada vez que leio algo à respeito destes dois status codes, fico mais na dúvida ainda... kkk

Reply:
> E na realidade, cada vez que leio algo à respeito destes dois status codes, fico mais na dúvida ainda...

O que me ajuda é ler não a diferença entre 204 e 404, mas sobre o significado do 2xx e do 4xx. Segundo a RFC em questão (tradução improvisada aqui):

> A classe de status 2xx (Sucesso) indica que a requisição do cliente foi recebida, compreendida e aceita com sucesso.
> (…)
> A classe de status 4xx (Erro do cliente) indica que o cliente parece ter se equivocado.

Com isso, imagino que a questão seja semântica e subjetiva (por isso evito fazer o @murilohns fez de _rotular_ como equívoco ou acerto). Ao meu ver, uma requisição com um CNPJ válido é uma requisição válida e, portanto, não deveria retornar 4xx.

Entendo que o _comum_ é o outro entendimento. Entendo que _geralmente_ usamos 404 — até pq usando 204 não podemos retornar conteúdo (e seria muito complicado, por exemplo o GitHub retorna 204 quando digitamos um nome de usuário ou repositório inexistente, ou mesmo de um repositório privado ao qual não temos acesso; pois o navegador mostraria uma página em branco). Mas como a API é voltada para o consumo por aplicação, com uma linguagem técnica, acredito que avisar que a requisição está correta (2xx) mas não há conteúdo (204) é uma boa alternativa.

**Mas, voltando ao ponto…** essa _branch_ no estado em que está, como podemos melhorar pensando no BrasilAPI?

Reply:
Cara, eu acho que a PR está legal do jeito que está, só não sei se vão fazer o merge da PR que implementa um middleware para cors e cache, talvez tenha que fazer alguns ajustes neste sentido se eles fizerem o merge antes :-p

Reply:
> só não sei se vão fazer o merge da PR que implementa um middleware para cors e cache

Massa! Se precisar mudar algo, só me avisarem : )